tests/suites/test_suite_ssl.function

```diff
@@ -797,9 +797,9 @@ int mbedtls_mock_tcp_recv_msg( void *ctx, unsigned char *buf, size_t buf_len )
  */
 typedef struct mbedtls_endpoint_certificate
 {
-    mbedtls_x509_crt ca_cert;
-    mbedtls_x509_crt cert;
-    mbedtls_pk_context pkey;
+    mbedtls_x509_crt* ca_cert;
```

Comment:
> The reason for this fix to be included here is that with so many failures due to unmet dependencies, it was an often occurence that certificate loading was failing, then, a partially setup structure was freed, and there was a segfault.

Ok, makes sense, that's the connection I was missing. And indeed I imagine it was painful to track, and clearly we don't want that to happen again.

> one of the certificates was freed twice

What's surprising to me is that calling `mbedtls_x509_crt_free()` twice in a row on the same object is supposed to work. The only circumstance where calling `mbedtls_x509_crt_free()` is not safe is when `mbedtls_x509_crt_init()` has not been called yet - but usually that can be avoided by placing calls to `init()` early enough in each function, before anything can fail. (We already had an issue where a test suite would coredump in some circumstance due to a late `init()`, fixed in #6109.)

Do you think it's possible that the root cause would be similar here?

Reply:
I'll sit down to investigate and describe it in more detail once I get the fixes in and the CI is busy :)

Reply:
OK, the basic issue was similar to what you described:
There are two endpoints, and their `mbedtls_endpoint_init()` were called in succession.
If the first call failed (because it not only prepares the structure, but also tries to parse certificates, etc) - the second one was not zeroized and it had some random data inside endpoint->cert (mbedtls_endpoint_certificate). Then, freeing this data would crash and burn.
I can recall however, that even the zeroization didn't solve all the problems, although I can't reproduce it locally (maybe it had something to do with the config), so I'll just create a partial, temporary PR to see if the CI helps me with my memory - https://github.com/Mbed-TLS/mbedtls/pull/6423 :)

Reply:
> If the first call failed (because it not only prepares the structure, but also tries to parse certificates, etc)

Ah, that's a bad pattern, we should have `mbedtls_endpoint_init()` that can't fail and just makes it safe to call `xxx_free()`, and then `mbedtls_endpoint_setup()` which actually populates the structure and may fail. (Or in a better world where the X.509 / TLS APIs would be more like the PSA APIs, we'd have `INIT` macros for everything so that we can initialize variables while declaring them.)

Though, an error can hide another, so indeed perhaps this is not the end of the story, let's see if the CI runs finds something else. But regardless, I think we should fix that bad pattern - if only to make sure we don't imitate it in other places.